util/fibers/uring_proactor.cc

```diff
@@ -554,27 +565,36 @@ void UringProactor::EpollDel(EpollIndex id) {
   uint64_t uid = entry->index;
   int fd = entry->fd;
 
-  if (sync_cancel_f_) {
-    io_uring_sync_cancel_reg reg_arg;
-    memset(&reg_arg, 0, sizeof(reg_arg));
-    reg_arg.timeout.tv_nsec = -1;
-    reg_arg.timeout.tv_sec = -1;
-    reg_arg.flags = 0;
-    reg_arg.addr = uid;
+  // using io_uring_register_sync_cancel is not applicable here because of several reasons:
+  // 1. it can race with EpollAdd requests that has not been processed yet, which will result
+  //    with missing cancellation.
+  // 2. It may race with callbacks that are already being in the completion queue.
+  //    Imagine for example the peer shuts down the socket and we have a pending POLLHUP
+  //    notification. io_uring_register_sync_cancel won't cancel the notification that is already
+  //    in the queue.
+  // Both scenarios lead to data race and crash as we delete the "entry" object at the end of
+  // the call.
+
+  // When using async POLL_REMOVE we **try** to guarantee serialized order.
+  // The assumption is that callback will always run after issuing PrepPollRemove.
+  IoResult res;
+  pending_epoll_removals_++;
+
+  do {
+    FiberCall fc(this);
+    fc->PrepPollRemove(uid);
```

Comment:
Ok after searching about this a little bit more it appears that only `io_uring_register_sync_cancel` can create orphan entries. We should not reach that state with PrepPollRemove as any other completion event related to that poll will be processed `before` the poll cancellation